pkg/analyzer/detection/detection.go

```diff
@@ -25,40 +26,63 @@ func DetectComponents(name string, files analyzer.Files, parent *database.Layer,
 	// If no CPEs are found on RHCOS, a note needs to be added that informs users of it.
 	// Also, the bool logic can be refactored into a single UseCertifiedWorkflow function.
 	// TODO(ROX-13906, ROX-14028): Implement note and refactor
-	if namespace != nil && (wellknownnamespaces.IsRHELNamespace(namespace.Name) || wellknownnamespaces.IsRHCOSNamespace(namespace.Name)) {
-		// This is a certified image that needs to be scanned differently.
-		// Use the RHELv2 scanner instead.
-		packages, cpes, err := rpm.ListFeatures(files)
+	if isCertifiedRHELNamespace(namespace) {
+		rhelFeatures, languageComponents, err = detectAndAnnotateCertifiedRHELComponents(name, files, namespace, languageComponents)
 		if err != nil {
 			return nil, nil, nil, nil, err
 		}
-		rhelfeatures = &database.RHELv2Components{
-			Dist:     namespace.Name,
-			Packages: packages,
-			CPEs:     cpes,
-		}
-		logrus.WithFields(logrus.Fields{LogLayerName: name, "rhel package count": len(packages), "rhel cpe count": len(cpes)}).Debug("detected rhelv2 features")
-		if err := rpm.AnnotateComponentsWithPackageManagerInfo(files, languageComponents); err != nil {
-			logrus.WithError(err).Errorf("Failed to analyze package manager info for language components: %s", name)
-		}
+		logrus.WithFields(logrus.Fields{
+			LogLayerName:         name,
+			"rhel package count": len(rhelFeatures.Packages),
+			"rhel cpe count":     len(rhelFeatures.CPEs),
+			"rhel content sets":  len(rhelFeatures.ContentSets),
+		}).Debug("detected rhelv2 features")
 	} else {
-		var err error
-		// Detect features.
 		featureVersions, err = detectFeatureVersions(name, files, namespace, parent)
 		if err != nil {
 			return nil, nil, nil, nil, err
 		}
-		if len(featureVersions) > 0 {
-			logrus.WithFields(logrus.Fields{LogLayerName: name, "feature count": len(featureVersions)}).Debug("detected features")
-		}
+		logrus.WithFields(logrus.Fields{
+			LogLayerName:    name,
+			"feature count": len(featureVersions),
+		}).Debug("detected features")
+	}
+	return namespace, featureVersions, rhelFeatures, languageComponents, nil
+}
+
+func isCertifiedRHELNamespace(namespace *database.Namespace) bool {
+	if namespace == nil {
+		return false
+	}
+	return wellknownnamespaces.IsRHELNamespace(namespace.Name) ||
+		wellknownnamespaces.IsRHCOSNamespace(namespace.Name)
+}
+
+func detectAndAnnotateCertifiedRHELComponents(name string, files analyzer.Files, namespace *database.Namespace, languageComponents []*component.Component) (*database.RHELv2Components, []*component.Component, error) {
+	// This is a certified image that needs to be scanned differently.
+	// Use the RHELv2 scanner instead.
+	packages, contentSets, err := rpm.ListFeatures(files)
+	if err != nil {
+		return nil, nil, err
+	}
+	rhelfeatures := &database.RHELv2Components{
+		Dist:     namespace.Name,
+		Packages: packages,
+		// CPEs are mapped and returned with content sets for backward compatibility.
+		CPEs:        repo2cpe.Singleton().Get(contentSets),
```

Comment:
Just realized this may panic if it attempts to load a file and fails. It won't panic as long as `REPO_TO_CPE_DIR` is not defined (or is empty). Maybe we should reconsider the panic. It made sense before Scanner started to be used as a library like this, but now if we mess up, Compliance pods can panic. At the moment, it won't, but it could in the future if we aren't careful